### ISSUE TITLE
Support for multipart form extractors

### ISSUE THREAD

**Issue by jacob-pro** [actix/actix-extras#278]
If you have a look on crates.io there are a proliferation of crates for doing multipart form uploads with files in actix. 

Including:

- [actix-easy-multipart](https://github.com/jacob-pro/actix-easy-multipart) by me
- [actix-multipart-extract](https://github.com/JSH32/actix-multipart-extract) by @JSH32
- [actix-multipart-derive](https://github.com/robjtede/actix-multipart-derive) by @robjtede 
- [actix-web-multipart-file](https://github.com/KeenS/actix-web-multipart-file) by @KeenS
- [[actix_extract_multipart](https://crates.io/crates/actix_extract_multipart)](https://github.com/KosekiDev/ActixExtractMultipart) by @KosekiDev
- [awmp](https://github.com/kardeiz/awmp) by @kardeiz
- [actix-form-data](https://crates.io/crates/actix-form-data) by @asonix

I think each of these have some unique improvements and features that the others do not. It seems silly to have so many unofficial crates, and is not very friendly to people who just want to quickly get started with actix.

I am wondering if we should have an effort to consolidate all of these crates into one official one in the actix project?

**Comment by jacob-pro** [comment]
Yep I think all of these support at least writing files. In my implementation stuff gets written to a [NamedTempfFile](https://docs.rs/tempfile/latest/tempfile/struct.NamedTempFile.html)  which you can then choose to persist. 

What I might do a bit later when I have time is write up a matrix of all the different features these crates have, so we can get an idea of what an official crate would need to include.

**Comment by asonix** [comment]
I also think that maybe moving to _just_ an Extractor would be a better idea than my current api, which is a combination Middleware + Extractor. Maybe I'll play with updating my library to work more like that

**Comment by JSH32** [comment]
Mine uses serde and does not support streaming. It was mostly just a proof of concept but I've been locally working on something similar to a builder but with a procmacro to auto parse into the struct.

**Comment by asonix** [comment]
@JSH32 fwiw the proc-macro "extract into this struct" method is really nice, and ideally that would be the final API. Maybe if there was an annotation you could put on a field like `#[stream_with(handler)]` to allow custom stream processing while still storing the result in the struct

**Comment by JSH32** [comment]
@asonix You could allow both `Vec<u8>` for getting just the bytes or something which is an `AsyncRead` (or something like that) to stream the bytes.

**Comment by jacob-pro** [comment]
@JSH32 yeah I was impressed with your serde compatible macro - it is a lot better than my one!

**Comment by asonix** [comment]
@JSH32 that doesn't quite work, because you need to process each field in order. If you embed a file stream directly into a struct for processing later, you don't read the Multipart stream any farther and can't "deserialize" fields that come afterwards.

**Comment by JSH32** [comment]
> @JSH32 that doesn't quite work, because you need to process each field in order. If you embed a file stream directly into a struct for processing later, you don't read the Multipart stream any farther and can't "deserialize" fields that come afterwards.

The problem is that a lot of the time the processor for the field will need to have context of the rest of the data, and reading the pure request is kind of unintuitive. Maybe save the original stream as a temporary file and provide a stream for reading the bytes from temporary? Or maybe optionally you can put any field which implements something like this?
```rs
pub trait ToStream {
    fn to_stream(field: actix_multipart::Field) -> StreamExt;
}
```

**Comment by asonix** [comment]
Saving to temporary files solves the problem for a number of use-cases, but prevents creating a pure "file proxy" implementation, which could stream files in-line to some other service. It requires either using tmpfs (which means holding files in memory) or using a filesystem, which adds requirements for deploying the service

I like the ToStream trait, but maybe it could look something like this:
```rust
trait FromStream {
    type Error: ResponseError;
    type Future: Future<Output = Result<Self, Self::Error>>,
    
    fn from_stream(req: &HttpRequest, stream: actix_multipart::Field /* or maybe some other opaque stream type */) -> Self::Future;
}

struct StreamMetadata {
    object_storage_id: String,
}

impl FromStream for StreamMetadata {
    type Error = actix_web::Error;
    type Future = Pin<Box<dyn Future<Output = Result<Self, Self::Error>>>>;
    
    fn from_stream(req: &HttpRequest, stream: Field) -> Self::Future {
        // extract useful things from HttpRequest here
        // ...
        
        Box::pin(async move {
            // upload to minio or s3 or something here
            // ...
            
            Ok(StreamMetadata { object_storage_id: /* blah */ })
        })
    }
}

#[derive(Multipart)]
struct MyForm {
    file_field: StreamMetadata
}
```

**Comment by JSH32** [comment]
That definitely works. Still kind of meh though because you cant get the full metadata of a request and will need to do processing of your form outside of the route but I'm pretty sure its the only way to do it right.

**Comment by asonix** [comment]
This `FromStream` trait could be implemented for `PathBuf` (or maybe a custom TemporaryPathbuf) that would stream the files onto a disk and return their path in the deserialized struct

**Comment by jacob-pro** [comment]
Thanks @asonix @JSH32 

I've had a think about it and I think these are all the features we will need:

#### What we need to do with streams of bytes:

1. Store the bytes in memory, e.g.
    ```rust
    pub struct BytePart {
        pub bytes: Vec<u8>,
        pub filename: Option<String>,
        pub mime: Mime,
    }
    ```
2. Temporary files on disk (that can later be persisted)
    ```rust
    pub struct FilePart {
        /// The file data itself stored as a temporary file on disk.
        pub file: NamedTempFile,
        /// The size in bytes of the file.
        pub size: usize,
        /// The `filename` value in the `Content-Disposition` header.
        pub filename: Option<String>,
        /// The Content-Type specified as reported in the uploaded form.
        pub mime: mime::Mime,
    }
    ```
3. An arbitrary user implemented future that streams the bytes elsewhere (e.g. cloud object storage)
   ```rust
   pub struct ObjectPart {
       /// The file has been uploaded to an external service
       pub object_id: String,
   }
    ```

As suggested by @asonix the solution could be to require each field to implement a trait, something like:

```rust
pub trait HandlePart {
   // Providing the request context allows pulling from app_data() to read config etc.
   fn handler(req: HttpRequest, stream: Field) -> Future<Output = Result<Self>>;
}

struct Form {
    bytes_in_memory: BytePart,
    file_on_disk: FilePart,
    file_in_s3: ObjectPart
}
```

#### What about simple text fields?

```rust
struct Form {
    text: String,
    a_number: i32,
    a_float: f64,
}
```

We could provide a handler for `T: FromStr` which would cover all of these?

#### How should we handle Lists and Optionals?

We would need to provide our own containers that implement the handler trait. (We can't implement directly on Vec/Option
because it would conflict with the generic FromStr implementation).

```rust
struct VecParts<T: HandlePart> (Vec<T>);
struct OptionalPart<T: HandlePart> (Option<T>);
```

According to the [spec](https://www.rfc-editor.org/rfc/rfc7578#section-4.3):

>    To match widely deployed implementations, multiple files MUST be sent
by supplying each file in a separate part but all with the same
"name" parameter.

#### Unknown and Extra fields

Serde allows using `#[serde(deny_unknown_fields)]` to prevent surplus fields - we should probably have something similar?

A further question is what to do if multiple fields are uploaded with the same name, but we are not expecting a Vector?

[My crate](https://github.com/jacob-pro/actix-easy-multipart/blob/master/actix-easy-multipart-derive/src/lib.rs#L11) 
has the `#[from_multipart(deny_extra_parts)]` which errors if:

- A form is uploaded with an unknown field name
- A form is uploaded with a duplicate field name (and the type is not a Vector)

But these two cases should really be separately configurable.

The duplicate field case should actually have 3 actions - keep first, keep last or error?

#### Renaming fields

We need to be able to support something similar to `#[serde(rename = "name")]` for cases where we can't use the field-name
e.g. rust keywords or different casing styles.

#### Data limits

We need to be able to set a global limit on the amount of data read into memory, for basic DDoS protection,
just as the official [actix extractors support](https://docs.rs/actix-web/latest/actix_web/web/struct.JsonConfig.html#method.limit).
It would have to be up to the implementor of the handler to choose whether to respect / consume from the in-memory limits.

But we will also want to provide field level limits as @JSH32 has done, which will be useful for files etc.

```rust
pub struct ExampleForm {
    #[multipart(max_size = 5MB)]
    file_field: File,
}
```



Let me know if you can think of anything I have missed?

**Comment by jacob-pro** [comment]
@robjtede Please can you advise if this sort of extractor would be accepted into acix-extras and what would be an appropriate name for it? 

Or actually maybe it should go into the [actix-multipart](https://github.com/actix/actix-web/tree/master/actix-multipart) crate itself - which would clear up a lot of the naming confusion?

If so, I am happy to write a lot of the code for this - but would just like to know where to get started from an organizational standpoint? Maybe we can have a WIP/feature branch in the actix repository which we can open PRs against?

**Comment by JSH32** [comment]
The handler will already be generated by a derive macro for introspecting field names/generics so Vec/Option can just be done there and conditionally code can be generated for that.

**Comment by jacob-pro** [comment]
@JSH32 I think you will find this is easier said than done? - I tried doing that with my macro

The problem is how do you tell if something is a `std::vec::Vec` or `std::option::Option` - I don't think macros are smart enough to do this - they can't do import resolution. E.g. what if a user is using a type definition etc.

**Comment by e-rhodes** [comment]
Is there a reason to use `FromStr` over an implementation with `serde` like @JSH32's? I may be misunderstanding the issue, but I'm imagining an interface like this:

```rust
/// The resulting struct from an upload.
/// 
/// All fields must either implement `Deserialize` or `HandlePart`
#[derive(Multipart)]
struct Upload {
    image: File,
    stream: Stream<Item = u8>,
    s3_upload: ImplementsS3Upload,
    id: Option<i32>,
    list: Vec<String>,
    metadata: Metadata,
}

#[derive(Deserialize)]
struct Metadata { ... }

struct ImplementsS3Upload { ... }
impl HandlePart for ImplementsS3Upload { ... }
```

**Comment by jacob-pro** [comment]
@e-rhodes Yeah I did also wonder about that - it would possibly make error handling easier - but my question is which deserializer were you thinking of using? 
We could use serde_plain - which works for primitive data types and simple enums.
https://github.com/mitsuhiko/serde-plain

**Comment by jacob-pro** [comment]
@e-rhodes I guess a further benefit if we use serde Deserialize as you suggest we could match on the provided mime type of the part, and use that to deserialize appropriately?
text/plain -> serde_plain
application/json -> serde_json
etc.

**Comment by e-rhodes** [comment]
I was mostly thinking about the additional form fields that can be posted with the files, as opposed to deserializing the content of the files. I don't have a ton of experience with the multipart spec, but it seems like other fields will just be `multipart/form-data` (per [this answer](https://stackoverflow.com/a/28380690/18145984)). 

For these fields, `serde_plain` may be sufficient---though it doesn't look like it would support deserializing multiple fields into a `#[serde(flatten)]`ed struct, or all the remaining keys into a `HashMap`, etc.

Looks like @JSH32's uses `serde_json`, but it supports the `field[]=item` syntax for arrays, which is awesome.

I also realized that my example above
```rust
    stream: Stream<Item = u8>,
    s3_upload: ImplementsS3Upload,
```
probably won't work because of the issue @asonix  and @JSH32  alluded to above, that you can't deserialize the rest of the fields without consuming the stream first. Would it be possible to successfully deserialize successfully only if the stream is provided last in the form? Would probably be confusing behavior to work with, but a nice interface if it works.

[The order is supposedly guaranteed](https://www.w3.org/TR/html4/interact/forms.html#h-17.13.4) but I'm not sure how that works with e.g. curl or other apis.

**Comment by jacob-pro** [comment]
@e-rhodes There are a few issues here:

First is the question of what syntax do you use for a list/array of parts. According [to the spec](https://www.rfc-editor.org/rfc/rfc7578#section-4.3) the recommended way to do it is that the uploader puts each part in with the same name. However this does vary based on implementation...

The reason I am suggesting having a separate `VecPart` and `OptionPart` type is because we should be able to upload a list of files (or an optional file) with an arbitrary handler:

```
#[derive(Multipart)]
struct Upload {
    files: VecPart<S3Upload>
    text: String,
}
```

So given the following upload:

```
Content-Type: multipart/form-data; boundary=BOUNDARY

--BOUNDARY
content-disposition: form-data; name="text"

Hello world
--BOUNDARY
content-disposition: form-data; name="files"

....bytes1...
--BOUNDARY
content-disposition: form-data; name="files"

....bytes2...
```

(Note the filename and content-type headers for each part are optional - it is up to the server to choose how to interpret them)

The multipart comes in as a stream of fields that we have to read one at a time. We would read the first part and see that its name is "text" we would then need to do a lookup (the macro would provide implementation of this lookup), that matches "text" -> a handler, in this case the one from `impl Handler for S3Upload` 

Once each handler is completed we will need to store its output in some sort of state, likely a Map<(String, TypeId), Box<Type> (similar concept to [actix extensions](https://docs.rs/actix-http/3.2.1/src/actix_http/extensions.rs.html#12).

Once we have processed all the parts in whatever order they arrived in, saved them into a map, we can then create the output struct `Upload` by retrieving each of its fields out of the map.

What we can't do is this:

```
#[derive(Multipart)]
struct Upload {
    files: Vec<S3Upload>
    text: String,
}
```

First of all S3Upload won't implement Deserialize, so Vec can't implement Deserialize. The compiler (at least until trait specialisation arrives) won't allow us to implement our handler trait for both `T: Deserialize` and also `Vec` at the same time, because of the possibility that `Vec` itself implements `Deserialize` (which it does!) causing a conflict .

We could compile this:

```
#[derive(Multipart)]
struct Upload {
    files: OptionalPart<S3Upload>
    text: Vec<String>,
    some_other_complex_structure: HashMap<String, String>,
    plain_text: String,
}
```

But it would only work where all the data for those fields is uploaded within a single part string. But to do that would require us to know which deserializer to use? One possibility is the user could provide it using the content-type field

```
Content-Type: multipart/form-data; boundary=BOUNDARY

--BOUNDARY
content-disposition: form-data; name="text"
content-type: application/json

["string 1", "string2"]
--BOUNDARY
content-disposition: form-data; name="some_other_complex_structure"
content-type: application/json

{"key": "value"}
--BOUNDARY
content-disposition: form-data; name="String"
content-type: text/plain

abcdefg
```

**Comment by jacob-pro** [comment]
Maybe https://github.com/robjtede/actix-web-lab is the place for development?

**Comment by e-rhodes** [comment]
> what syntax do you use for a list/array of parts

Seems like we should support both the multiple files with same field name as well as multiple fields with names like `field[]` (ideally for both the fields and the files).

> The compiler (at least until trait specialisation arrives) won't allow us to implement our handler trait for both T: Deserialize and also Vec at the same time, because of the possibility that Vec itself implements Deserialize (which it does!) causing a conflict

Makes sense now. Still, having wrappers for Vec and Option is less that ideal ux. Could we achieve this with attribute macros?

```rust
struct Upload {
    all_files: Vec<File>,
    cloud_file: S3Upload,
    #[multipart(deserialize)]
    field: Vec<String>,
}
```

Alternatively, maybe the library could provide a struct:
```rust
struct Multipart<F: HandlePart, T: Deserialize> {
    files: F,
    fields: T,
}
#[derive(HandlePart)]
struct Files { ... }
#[derive(Deserialize)]
struct Fields { ... }
pub async fn handler(upload: multipart::Multipart<Files, Fields>) { ... }
```

> the user could provide it using the content-type field

This seems like it would be frustrating to require, since [as far as I can tell](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) there's no way to supply content-type on an HTML form. I know you can do it with curl, but at least for my use case, I need a frontend. I think supporting deserialization of JSON passed as a form field is a low-priority use case.

For me, the most valuable use cases would just involve deserializing from flat form fields to enums (either validating a dropdown or fields that can be multiple types) or (not as useful) collecting multiple fields into a struct/map. E.g.:

```
Content-Type: multipart/form-data; boundary=BOUNDARY

--BOUNDARY
content-disposition: form-data; name="field1"

dog
--BOUNDARY
content-disposition: form-data; name="multi_type"

1
--BOUNDARY
content-disposition: form-data; name="other_field"

cat
--BOUNDARY
content-disposition: form-data; name="dropdown"

bird
--BOUNDARY
content-disposition: form-data; name="file"

## bytes of file ##
```


```rust
struct Data {
     field1: String
}
#[serde(untagged)]
enum StrOrInt {
    Str(String),
    Int(i32),
}
#[serde(rename_all = "snake_case")]
enum Dropdown {
    Bird,
    Frog,
}

#[derive(Multipart)]
struct Upload {
    #[serde(flatten)]
    data: Data,
    multi_type: StrOrInt, 
    dropdown: Dropdown,
    #[serde(flatten)]
    rest_of_the_fields: HashMap<String, Value>, // grabs other_field
    file: File,
}


```

**Comment by jacob-pro** [comment]
@e-rhodes 

> Seems like we should support both the multiple files with same field name as well as multiple fields with names like field[]

Do you mean where multiple parts are uploaded each using the name `field[]` ? Would the following work?

struct Upload {
    #[multipart(rename="files[]")]
    files: VecPart<File>,
    #[multipart(rename="text[]")]
    texts: VecPart<String>,
}

In the example you've given you wouldn't be able to use those `#[serde(flatten)]` annotations since as mentioned before the `Upload` struct itself does not and cannot implement `Deserialize` itself.

This also doesn't work because of the limitations of serde_plain:

```rust
    #[derive(Deserialize, Debug)]
    #[serde(untagged)]
    enum StrOrInt {
        Int(i32),
        Str(String),
    }

    #[test]
    fn test_deserialize() {
        let s1: StrOrInt = serde_plain::from_str("1234").unwrap();
        // Str("1234") - doesn't match to int because plain text doesn't make a distinction unlike JSON
        println!("{:?}", s1);
    }
```

But the other enum would work, and this is one advantage of implementing for `T: Deserialize` instead of just `T: FromStr`:

```
    #[derive(Deserialize, Debug)]
    #[serde(rename_all = "snake_case")]
    enum Dropdown {
        Bird,
        Frog,
    }

    #[test]
    fn test_deserialize2() {
        let d1: Dropdown = serde_plain::from_str("frog").unwrap();
        // Dropdown::Frog
        println!("{:?}", d1);
    }
```

I think the fundamental issue here is that you are trying to categorise things into files and non files...

From the perspective of the server we just get a stream of parts, each one has:

- A field name
- An optional filename (this doesn't necessarily indicate if it is a file)
- An optional mime type (but if it is missing, we can default to `text/plain`)
- A stream of bytes

>  The file name isn't mandatory
   for cases where the file name isn't available or is meaningless or
   private; this might result, for example, when selection or drag-and-
   drop is used or when the form data content is streamed directly from
   a device.

What the proposal above was going to do is map a handler based purely on the field name. If the field name is associated with the S3 uploader type, the data would get uploaded to S3 regardless of whether it is a 10 byte `text-plain` or a 1 GiB file `application/octet-stream` 

I guess a scenario to imagine is that you have a user uploading a 1GiB .txt file, it would have a mime type of `text/plain` and the filename may be empty - there is no reliable way for the server to decide if that is "file" or "non-file"...

> This seems like it would be frustrating to require, since there's no way to supply content-type on an HTML form.

Yes you are right, and it goes back to my question - that for a given part how do you know what deserializer to use?

My suggestion to use the content-type would work for HTML forms, since the forms either supply a mime type of `text/plain` or default to `text/plain` and therefore their fields would get deserialized with serde_plain.

If you wanted to do JSON, we would either need an API client to tell the server via the content type that it is uploading JSON within that part, or put in some sort of annotation on the server side so we know to deserialize that part with JSON instead.

> rest_of_the_fields: HashMap<String, T>, // grabs other_field

This is a feature I hadn't really thought about. It would be workable with an [anotation similar to this](https://github.com/serde-rs/serde/issues/941). But it would come with the restriction that every part would have to use the same handler `T`, e.g. you have a to choose whether you want to attempt to read every unknown field into memory, write to disk or upload to S3, etc.
